File: app/src/main/java/com/starostinvlad/fan/VideoScreen/VideoActivity.java
package com.starostinvlad.fan.VideoScreen;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.PictureInPictureParams;
import android.app.UiModeManager;
import android.content.Intent;
import android.content.res.Configuration;
import android.graphics.Color;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.media.session.MediaSessionCompat;
import android.util.Log;
import android.util.Rational;
import android.view.GestureDetector;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.appodeal.ads.Appodeal;
import com.google.android.exoplayer2.Format;
import com.google.android.exoplayer2.MediaItem;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.ext.mediasession.MediaSessionConnector;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.ui.AspectRatioFrameLayout;
import com.google.android.exoplayer2.ui.PlayerView;
import com.google.android.exoplayer2.util.MimeTypes;
import com.google.android.exoplayer2.util.Util;
import com.google.android.material.snackbar.Snackbar;
import com.starostinvlad.fan.Adapters.SubtitlesAdapter;
import com.starostinvlad.fan.Adapters.TranslationsAdapter;
import com.starostinvlad.fan.Adapters.SeasonRecyclerViewAdapter;
import com.starostinvlad.fan.App;
import com.starostinvlad.fan.GsonModels.News;
import com.starostinvlad.fan.R;
import com.starostinvlad.fan.VideoScreen.PlayerModel.Episode;
import com.starostinvlad.fan.VideoScreen.PlayerModel.Hls;
import com.starostinvlad.fan.VideoScreen.PlayerModel.Season;
import com.starostinvlad.fan.VideoScreen.PlayerModel.Serial;
import com.starostinvlad.fan.VideoScreen.PlayerModel.Translation;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GestureDetectorCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;


public class VideoActivity extends AppCompatActivity implements VideoActivityContract {
    private final String TAG = getClass().getSimpleName();
    int[] quality = {414000, 714000, 1064000, 5055521};

    PlayerView playerView;
    Toolbar toolbar;
    ProgressBar progressBar;
    View videoContainer;
    private Window window;
    private SimpleExoPlayer player;
    private VideoPresenter videoPresenter;
    private long currentPosition = 0;
    private boolean isInPictureInPictureMode;
    private DefaultTrackSelector trackSelector;
    private UiModeManager uiModeManager;
    private boolean viewed;
    private boolean subscribed;
    private String id = "";
    private Button exo_rew;
    private Button exo_ffwd;
    private ImageButton subtitlesBtn;
    private boolean doubleTap;
    private View doubleClickArea;


    public static void start(Activity activity, Serial serial) {
        Intent intent = new Intent(activity, VideoActivity.class);
        if (serial != null)
            intent.putExtra("SERIAL", serial);
        activity.startActivity(intent);
    }


    @SuppressLint("ClickableViewAccessibility")
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fragment_video);
        playerView = findViewById(R.id.player_view_id);
        toolbar = findViewById(R.id.video_fragment_toolbar);
        progressBar = findViewById(R.id.video_progress_id);
        videoContainer = findViewById(R.id.video_container);

        mDetector = new GestureDetectorCompat(this, new MyGestureListener());

//        if (getIntent() != null && getIntent().hasExtra(getString(R.string.episode_extra)))
//            episode = (News) getIntent().getSerializableExtra(getString(R.string.episode_extra));
//        else if (getIntent() != null && getIntent().hasExtra("NAME")) {
//            episode = new News();
//            episode.setTitle(getIntent().getStringExtra("NAME"));
//            episode.setHref(getIntent().getStringExtra("HREF"));
//        } else {
//            showDialog(getString(R.string.video_error));
//        }

        window = this.getWindow();
        uiModeManager = (UiModeManager) getSystemService(UI_MODE_SERVICE);
        hideSystemsElements();
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }


//        Log.d(TAG, "episode: " + episode.getSubTitle());

        trackSelector = new DefaultTrackSelector();
        DefaultTrackSelector.Parameters defaultTrackParam = trackSelector.buildUponParameters().build();
        trackSelector.setParameters(defaultTrackParam);

        playerView.setControllerShowTimeoutMs(2000);
        playerView.setControllerAutoShow(true);
        playerView.hideController();

        playerView.setShowBuffering(PlayerView.SHOW_BUFFERING_ALWAYS);
        player = new SimpleExoPlayer.Builder(this).setTrackSelector(trackSelector).build();
        playerView.setPlayer(player);

        exo_ffwd = findViewById(R.id.exo_ffwd);
        exo_rew = findViewById(R.id.exo_rew);

        ImageButton voices = findViewById(R.id.voice_btn);
        subtitlesBtn = findViewById(R.id.btn_subtitle);
        subtitlesBtn.setOnClickListener(view -> subtitleSelectorDialog());

        voices.setOnClickListener(view -> videoPresenter.onBuildDialog());

        doubleClickArea = findViewById(R.id.doubleClickArea);

        videoPresenter = new VideoPresenter();
        videoPresenter.attachView(this);

        if (getIntent() != null && getIntent().hasExtra("SERIAL")) {
            videoPresenter.onStartWithSerial((Serial) getIntent().getSerializableExtra("SERIAL"));
        } else if (App.getInstance().isReview()) {
            playerView.hideController();
            showDialog(getString(R.string.on_review));
        } else {
            Appodeal.show(this, Appodeal.INTERSTITIAL);
//            videoPresenter.loadData(episode.getHref());
        }
    }

    @Override
    public void showDialog(String msg) {
        new AlertDialog.Builder(this)
                .setMessage(msg)
                .setPositiveButton("OK", (dialogInterface, i) -> dialogInterface.dismiss())
                .setOnDismissListener(dialogInterface -> finish()).show();
    }

    @Override
    public void openTrailer(String url) {
        Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
        startActivity(browserIntent);
    }

    @Override
    public void changeDescription(String title, String subTitle) {
        toolbar.setTitle(title);
        toolbar.setSubtitle(subTitle);
        setSupportActionBar(toolbar);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_DPAD_CENTER) {
            if (!playerView.isControllerVisible())
                playerView.showController();
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public void initRecycle(List<Episode> episodes) {
        RecyclerView recyclerView = findViewById(R.id.episodeListRV);
        try {
            LinearLayoutManager horizontalLayoutManager
                    = new LinearLayoutManager(VideoActivity.this, LinearLayoutManager.HORIZONTAL, false);
            recyclerView.setLayoutManager(horizontalLayoutManager);
            SeasonRecyclerViewAdapter adapter = new SeasonRecyclerViewAdapter(this, episodes);
            adapter.setClickListener((view, position) -> {
                videoPresenter.onChangeEpisode(position);
                Appodeal.show(this, Appodeal.INTERSTITIAL);
            });
            recyclerView.setAdapter(adapter);
            Log.d(TAG, "episodes: " + episodes);
        } catch (Exception e) {
            e.printStackTrace();
        }
        Log.d(TAG, "fill complete");
    }

    @Override
    public void translationSelectorDialog(List<Translation> translations) {
        TranslationsAdapter adapter = new TranslationsAdapter(translations);
        AlertDialog.Builder alertBuilder = new AlertDialog.Builder(this)
                .setAdapter(adapter, (dialogInterface, i) -> videoPresenter.onChangeTranslation(i));
        alertBuilder.show();
    }

    @Override
    public void qualitySelectorDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(getString(R.string.quality))
                .setItems(R.array.quality_array, (dialog, which) -> {
                    DefaultTrackSelector.Parameters
                            parameters = trackSelector.buildUponParameters()
                            .setMaxVideoBitrate(quality[which])
                            .setForceHighestSupportedBitrate(true)
                            .build();
                    trackSelector.setParameters(parameters);
                });
        builder.show();
    }

    void subtitleSelectorDialog() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            SubtitlesAdapter adapter = new SubtitlesAdapter(player.getCurrentMediaItem().playbackProperties.subtitles);
            AlertDialog.Builder alertBuilder = new AlertDialog.Builder(this)
                    .setAdapter(adapter, (dialogInterface, i) -> {
                        Log.d(TAG, "showSubtitleDialog: lang: " + adapter.getItem(i));
                        DefaultTrackSelector.Parameters
                                parameters = trackSelector.buildUponParameters()
                                .setPreferredTextLanguage(adapter.getItem(i))
                                .build();
                        trackSelector.setParameters(parameters);
                    });
            alertBuilder.show();
        }
    }


    private void hideSystemsElements() {
        window.getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LOW_PROFILE
                | View.SYSTEM_UI_FLAG_FULLSCREEN
                | View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION);
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);
        window.setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
                WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.video, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
            case R.id.pip:
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O && uiModeManager.getCurrentModeType() != Configuration.UI_MODE_TYPE_TELEVISION) {
                    currentPosition = player.getCurrentPosition();
                    PictureInPictureParams pictureInPictureParams = new PictureInPictureParams.Builder()
                            .setAspectRatio(new Rational(16, 9))
                            .build();
                    enterPictureInPictureMode(pictureInPictureParams);
                    playerView.hideController();
                    player.play();
                }
                Log.i(TAG, "id " + item.getItemId());
                return true;

            case R.id.quality:
                qualitySelectorDialog();
                Log.i(TAG, "id " + item.getItemId());
                return true;
            case R.id.crop_video:
                if (playerView.getResizeMode() == AspectRatioFrameLayout.RESIZE_MODE_FIT)
                    playerView.setResizeMode(AspectRatioFrameLayout.RESIZE_MODE_ZOOM);
                else
                    playerView.setResizeMode(AspectRatioFrameLayout.RESIZE_MODE_FIT);
                Log.i(TAG, "id " + item.getItemId());
                return true;
        }
        return false;
    }

    @Override
    public void onPictureInPictureModeChanged(boolean isInPictureInPictureMode, Configuration newConfig) {
        Log.d(TAG, "OnPip mode");
        this.isInPictureInPictureMode = isInPictureInPictureMode;
//        if (player != null)
//            player.seekTo(currentPosition);
    }

    @Override
    public void initPlayer(Hls hls) {
        if (player == null) {
            player = new SimpleExoPlayer.Builder(this).setTrackSelector(trackSelector).build();
            playerView.setPlayer(player);
        }
        Log.d(TAG, "uri: " + hls.src);
        List<MediaItem.Subtitle> subtitles = new ArrayList<>();
        if (!hls.enSub.isEmpty() && !hls.enSub.equals("false")) {
            Uri sub = Uri.parse(hls.enSub);
            MediaItem.Subtitle subtitle = new MediaItem.Subtitle(sub, MimeTypes.TEXT_VTT, "en", Format.NO_VALUE);
            subtitles.add(subtitle);
        }
        if (!hls.ruSub.isEmpty() && !hls.ruSub.equals("false")) {
            Uri sub = Uri.parse(hls.ruSub);
            MediaItem.Subtitle subtitle = new MediaItem.Subtitle(sub, MimeTypes.TEXT_VTT, "ru", Format.NO_VALUE);
            subtitles.add(subtitle);
        }
        if (subtitles.isEmpty()) {
            subtitlesBtn.setVisibility(View.GONE);
        } else {
            subtitlesBtn.setVisibility(View.VISIBLE);
        }
        MediaItem mediaItem = new MediaItem.Builder()
                .setUri(hls.src)
                .setSubtitles(subtitles)
                .build();
//        MediaItem mediaItem = MediaItem.fromUri(uri);
        Log.d(TAG, "media item: " + mediaItem.playbackProperties.uri);
//        player = new SimpleExoPlayer.Builder(this).build();
        if (player != null) {
            player.setMediaItem(mediaItem);
            player.setForegroundMode(true);
//        playerView.setPlayer(player);
            player.prepare();
            player.seekTo(currentPosition);
        }
    }

    @Override
    public void alarm(String message) {
        Snackbar.make(videoContainer, message, Snackbar.LENGTH_LONG)
                .setBackgroundTint(Color.RED)
                .show();
    }

    private void releasePlayer() {
        if (player != null) {
            player.release();
            player = null;
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        videoPresenter.onStart();
        MediaSessionCompat mediaSession = new MediaSessionCompat(this,getPackageName());
        MediaSessionConnector mediaSessionConnector = new MediaSessionConnector(mediaSession);
        mediaSessionConnector.setPlayer(player);
        mediaSession.setActive(true);
    }

    @Override
    public void onPause() {
        super.onPause();
        if (!isInPictureInPictureMode) {
            if (player != null) {
                currentPosition = player.getCurrentPosition();
            }
            if (Util.SDK_INT <= 23) {
                releasePlayer();
            }
        }

    }

    @Override
    public void onStop() {
        super.onStop();
        if (Util.SDK_INT > 23) {
            releasePlayer();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        videoPresenter.onDetach();
    }

    @Override
    public void showLoading(boolean b) {
        videoContainer.setVisibility(!b ? View.VISIBLE : View.GONE);
        toolbar.setVisibility(!b ? View.VISIBLE : View.GONE);
        progressBar.setVisibility(b ? View.VISIBLE : View.GONE);
    }

    @Override
    public void fillToolbar(String title) {
        toolbar.setTitle(title);
        setSupportActionBar(toolbar);
    }


    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        Log.d(TAG, "dispatchTouchEvent: dispatch touch");
        if (uiModeManager.getCurrentModeType() != Configuration.UI_MODE_TYPE_TELEVISION)
            this.mDetector.onTouchEvent(event);
        Log.d(TAG, "dispatchTouchEvent: doubleTap 1:" + doubleTap);
        if (!doubleTap) {
            hideSystemsElements();
            return playerView.dispatchTouchEvent(event);
        }
        doubleTap = false;
//        else
//            return mDetector.onTouchEvent(event);
//            return true;

        return super.dispatchTouchEvent(event);
    }

    GestureDetectorCompat mDetector;

    class MyGestureListener extends GestureDetector.SimpleOnGestureListener {
        private static final String DEBUG_TAG = "Gestures";

        @Override
        public boolean onDown(MotionEvent event) {
            Log.d(DEBUG_TAG, "onDown: " + event.toString());
            return true;
        }

        @Override
        public boolean onDoubleTapEvent(MotionEvent e) {
            doubleTap = true;
            return super.onDoubleTapEvent(e);
        }

        @Override
        public boolean onDoubleTap(MotionEvent event) {
            Log.d(DEBUG_TAG, "onDoubleTapEvent: tap " + event.getRawX() + " : " + playerView.getWidth() / 2.0);
            doubleTap = false;
            doubleClickArea.performClick();
            if (event.getRawX() > playerView.getWidth() / 2.0) {
                exo_ffwd.performClick();
            } else {
                exo_rew.performClick();
            }
            return super.onDoubleTap(event);
        }
    }
}

File: app/src/main/java/com/starostinvlad/fan/VideoScreen/VideoActivity.kt
package com.starostinvlad.fan.VideoScreen

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Dialog
import android.app.PictureInPictureParams
import android.app.UiModeManager
import android.content.DialogInterface
import android.content.Intent
import android.content.res.Configuration
import android.graphics.Color
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.support.v4.media.session.MediaSessionCompat
import android.util.Log
import android.util.Rational
import android.view.*
import android.view.GestureDetector.SimpleOnGestureListener
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageButton
import android.widget.ProgressBar
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.GestureDetectorCompat
import androidx.core.view.isVisible
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.appodeal.ads.Appodeal
import com.google.android.exoplayer2.Format
import com.google.android.exoplayer2.MediaItem
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.ext.mediasession.MediaSessionConnector
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import com.google.android.exoplayer2.ui.AspectRatioFrameLayout
import com.google.android.exoplayer2.ui.PlayerView
import com.google.android.exoplayer2.util.MimeTypes
import com.google.android.exoplayer2.util.Util
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.google.android.material.snackbar.Snackbar
import com.starostinvlad.fan.Adapters.DialogRVAdapter
import com.starostinvlad.fan.Adapters.OnItemSelectListener
import com.starostinvlad.fan.Adapters.SeasonRecyclerViewAdapter
import com.starostinvlad.fan.App
import com.starostinvlad.fan.R
import com.starostinvlad.fan.VideoScreen.PlayerModel.Episode
import com.starostinvlad.fan.VideoScreen.PlayerModel.Hls
import com.starostinvlad.fan.VideoScreen.PlayerModel.Serial
import com.starostinvlad.fan.VideoScreen.PlayerModel.Translation

class VideoActivity : AppCompatActivity(), VideoActivityContract {
    private val TAG: String = this::class.simpleName!!
    var quality = mapOf(
            "360" to 414000,
            "480" to 714000,
            "720" to 1064000,
            "1080" to 5055521
    )
    lateinit var playerView: PlayerView
    private var toolbar: Toolbar? = null
    private var progressBar: ProgressBar? = null
    private var videoContainer: View? = null
    private var simpleExoPlayer: SimpleExoPlayer? = null
    private var videoPresenter: VideoPresenter? = null
    private var currentPosition: Long = 0
    private var isInPIPMode = false
    private var trackSelector: DefaultTrackSelector? = null
    private var uiModeManager: UiModeManager? = null
    private var exoRew: Button? = null
    private var exoFfwd: Button? = null
    private var subtitlesBtn: ImageButton? = null
    private var doubleTap = false
    private var doubleClickArea: View? = null
    private var mDetector: GestureDetectorCompat? = null


    @SuppressLint("ClickableViewAccessibility")
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.fragment_video)
        playerView = findViewById(R.id.player_view_id)
        toolbar = findViewById(R.id.video_fragment_toolbar)
        progressBar = findViewById(R.id.video_progress_id)
        videoContainer = findViewById(R.id.video_container)
        mDetector = GestureDetectorCompat(this, MyGestureListener())

//        if (getIntent() != null && getIntent().hasExtra(getString(R.string.episode_extra)))
//            episode = (News) getIntent().getSerializableExtra(getString(R.string.episode_extra));
//        else if (getIntent() != null && getIntent().hasExtra("NAME")) {
//            episode = new News();
//            episode.setTitle(getIntent().getStringExtra("NAME"));
//            episode.setHref(getIntent().getStringExtra("HREF"));
//        } else {
//            showDialog(getString(R.string.video_error));
//        }
        uiModeManager = getSystemService(UI_MODE_SERVICE) as UiModeManager
        hideSystemsElements()
        setSupportActionBar(toolbar)
        if (supportActionBar != null) {
            supportActionBar!!.setHomeButtonEnabled(true)
            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        }


//        Log.d(TAG, "episode: " + episode.getSubTitle());
        trackSelector = DefaultTrackSelector(this)
        val defaultTrackParam = trackSelector!!.buildUponParameters().build()
        trackSelector!!.parameters = defaultTrackParam
        with(playerView) {
            controllerShowTimeoutMs = 2000
            controllerAutoShow = true
            hideController()
            setShowBuffering(PlayerView.SHOW_BUFFERING_ALWAYS)
            simpleExoPlayer = SimpleExoPlayer.Builder(context).setTrackSelector(trackSelector!!).build()
            player = simpleExoPlayer
        }


        exoFfwd = findViewById(R.id.exo_ffwd)
        exoRew = findViewById(R.id.exo_rew)
        val voices = findViewById<ImageButton>(R.id.voice_btn)
        subtitlesBtn = findViewById(R.id.btn_subtitle)
        subtitlesBtn?.setOnClickListener { subtitleSelectorDialog() }
        voices.setOnClickListener { videoPresenter!!.onBuildDialog() }
        doubleClickArea = findViewById(R.id.doubleClickArea)
        videoPresenter = VideoPresenter()
        videoPresenter!!.attachView(this)
        if (intent != null && intent.hasExtra("SERIAL")) {
            videoPresenter!!.onStartWithSerial(intent.getSerializableExtra("SERIAL") as Serial)
        } else if (App.instance.isReview) {
            playerView.hideController()
            showDialog(getString(R.string.on_review))
        } else {
            Appodeal.show(this, Appodeal.INTERSTITIAL)
            //            videoPresenter.loadData(episode.getHref());
        }
    }

    override fun showDialog(msg: String) {
        AlertDialog.Builder(this)
                .setMessage(msg)
                .setPositiveButton("OK") { dialogInterface: DialogInterface, _: Int -> dialogInterface.dismiss() }
                .setOnDismissListener { finish() }.show()
    }

    override fun openTrailer(url: String) {
        val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        startActivity(browserIntent)
    }

    override fun changeDescription(title: String, subTitle: String) {
        toolbar!!.title = title
        toolbar!!.subtitle = subTitle
        setSupportActionBar(toolbar)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_DPAD_CENTER) {
            if (!playerView.isControllerVisible) playerView.showController()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun initRecycle(episodes: List<Episode>) {
        val recyclerView = findViewById<RecyclerView>(R.id.episodeListRV)
        try {
            val horizontalLayoutManager = LinearLayoutManager(this@VideoActivity, LinearLayoutManager.HORIZONTAL, false)
            recyclerView.layoutManager = horizontalLayoutManager
            val adapter = SeasonRecyclerViewAdapter()
            adapter.items = episodes.toMutableList()
            adapter.mClickListener = { it ->
                videoPresenter!!.onChangeEpisode(it)
                Appodeal.show(this, Appodeal.INTERSTITIAL)
            }
            recyclerView.adapter = adapter
            Log.d(TAG, "episodes: $episodes")
        } catch (e: Exception) {
            e.printStackTrace()
        }
        Log.d(TAG, "fill complete")
    }

    override fun translationSelectorDialog(translations: List<Translation>) {
        val dialogFragment = DialogFragment(items = translations.map { translation -> translation.title }, callback = object : DialogCallback {
            override fun onSelect(position: Int) {
                videoPresenter!!.onChangeTranslation(position)
            }
        })
        dialogFragment.show(supportFragmentManager, "Озвучка")
    }

    override fun qualitySelectorDialog() {
        val dialogFragment =
                DialogFragment(items = quality.keys.toList(),
                        callback = object : DialogCallback {
                            override fun onSelect(position: Int) {
                                val parameters = trackSelector!!.buildUponParameters()
                                        .setMaxVideoBitrate(quality.getValue(quality.keys.toList()[position]))
                                        .setForceHighestSupportedBitrate(true)
                                        .build()
                                trackSelector!!.parameters = parameters
                            }
                        })
        dialogFragment.show(supportFragmentManager, "Качество")
    }

    private fun subtitleSelectorDialog() {
        val subs = simpleExoPlayer!!
                .currentMediaItem!!
                .playbackProperties!!
                .subtitles
                .map { it.language!! }

        val dialogFragment = DialogFragment(items = subs, callback = object : DialogCallback {
            override fun onSelect(position: Int) {
                val parameters = trackSelector!!.buildUponParameters()
                        .setPreferredTextLanguage(
                                subs[position]
                        )
                        .build()
                trackSelector!!.parameters = parameters
            }
        })
        dialogFragment.show(supportFragmentManager, "Субтитры")
    }

    private fun hideSystemsElements() {
        with(window) {
            decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LOW_PROFILE
                    or View.SYSTEM_UI_FLAG_FULLSCREEN
                    or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                    or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION)
            setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                    WindowManager.LayoutParams.FLAG_FULLSCREEN)
            setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
                    WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
        }

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.video, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
            R.id.pip -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && uiModeManager!!.currentModeType != Configuration.UI_MODE_TYPE_TELEVISION) {
                    currentPosition = simpleExoPlayer!!.currentPosition
                    val pictureInPictureParams = PictureInPictureParams.Builder()
                            .setAspectRatio(Rational(16, 9))
                            .build()
                    enterPictureInPictureMode(pictureInPictureParams)
                    playerView.hideController()
                    simpleExoPlayer!!.play()
                }
                Log.i(TAG, "id " + item.itemId)
                return true
            }
            R.id.quality -> {
                qualitySelectorDialog()
                Log.i(TAG, "id " + item.itemId)
                return true
            }
            R.id.crop_video -> {
                if (playerView.resizeMode == AspectRatioFrameLayout.RESIZE_MODE_FIT) playerView.resizeMode = AspectRatioFrameLayout.RESIZE_MODE_ZOOM else playerView.resizeMode = AspectRatioFrameLayout.RESIZE_MODE_FIT
                Log.i(TAG, "id " + item.itemId)
                return true
            }
        }
        return false
    }

    override fun onPictureInPictureModeChanged(isInPictureInPictureMode: Boolean, newConfig: Configuration) {
        Log.d(TAG, "OnPip mode")
        isInPIPMode = isInPictureInPictureMode
        //        if (player != null)
//            player.seekTo(currentPosition);
    }

    override fun initPlayer(hls: Hls) {
        if (simpleExoPlayer == null) {
            simpleExoPlayer = SimpleExoPlayer.Builder(this).setTrackSelector(trackSelector!!).build()
            playerView.player = simpleExoPlayer
        }
        Log.d(TAG, "uri: " + hls.src)
        val subtitles: MutableList<MediaItem.Subtitle> = ArrayList()
        if (hls.enSub.isNotEmpty() && hls.enSub != "false") {
            val sub = Uri.parse(hls.enSub)
            val subtitle = MediaItem.Subtitle(sub, MimeTypes.TEXT_VTT, "en", Format.NO_VALUE)
            subtitles.add(subtitle)
        }
        if (hls.ruSub.isNotEmpty() && hls.ruSub != "false") {
            val sub = Uri.parse(hls.ruSub)
            val subtitle = MediaItem.Subtitle(sub, MimeTypes.TEXT_VTT, "ru", Format.NO_VALUE)
            subtitles.add(subtitle)
        }
        if (subtitles.isEmpty()) {
            subtitlesBtn!!.visibility = View.GONE
        } else {
            subtitlesBtn!!.visibility = View.VISIBLE
        }
        val mediaItem = MediaItem.Builder()
                .setUri(hls.src)
                .setSubtitles(subtitles)
                .build()
        Log.d(TAG, "media item: " + mediaItem.playbackProperties!!.uri)
        if (simpleExoPlayer != null) {
            with(simpleExoPlayer!!) {
                setMediaItem(mediaItem)
                setForegroundMode(true)
                prepare()
                seekTo(currentPosition)
            }
        }
    }

    override fun alarm(message: String) {
        Snackbar.make(videoContainer!!, message, Snackbar.LENGTH_LONG)
                .setBackgroundTint(Color.RED)
                .show()
    }

    private fun releasePlayer() {
        if (simpleExoPlayer != null) {
            simpleExoPlayer!!.release()
            simpleExoPlayer = null
        }
    }

    public override fun onStart() {
        super.onStart()
        videoPresenter!!.onStart()
        val mediaSession = MediaSessionCompat(this, packageName)
        val mediaSessionConnector = MediaSessionConnector(mediaSession)
        mediaSessionConnector.setPlayer(simpleExoPlayer)
        mediaSession.isActive = true
    }

    public override fun onPause() {
        super.onPause()
        if (!isInPIPMode) {
            if (simpleExoPlayer != null) {
                currentPosition = simpleExoPlayer!!.currentPosition
            }
            if (Util.SDK_INT <= 23) {
                releasePlayer()
            }
        }
    }

    public override fun onStop() {
        super.onStop()
        if (Util.SDK_INT > 23) {
            releasePlayer()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        videoPresenter!!.onDetach()
    }

    override fun showLoading(show: Boolean) {
        videoContainer!!.isVisible = !show
        toolbar!!.isVisible = !show
        progressBar!!.isVisible = show
    }

    override fun fillToolbar(title: String) {
        toolbar!!.title = title
        setSupportActionBar(toolbar)
    }

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        Log.d(TAG, "dispatchTouchEvent: dispatch touch")
        if (uiModeManager!!.currentModeType != Configuration.UI_MODE_TYPE_TELEVISION) mDetector!!.onTouchEvent(event)
        Log.d(TAG, "dispatchTouchEvent: doubleTap 1:$doubleTap")
        if (!doubleTap) {
            hideSystemsElements()
            return playerView.dispatchTouchEvent(event)
        }
        doubleTap = false
        //        else
//            return mDetector.onTouchEvent(event);
//            return true;
        return super.dispatchTouchEvent(event)
    }

    internal inner class MyGestureListener : SimpleOnGestureListener() {
        override fun onDown(event: MotionEvent): Boolean {
            Log.d(TAG, "onDown: $event")
            return true
        }

        override fun onDoubleTapEvent(e: MotionEvent): Boolean {
            doubleTap = true
            return super.onDoubleTapEvent(e)
        }

        override fun onDoubleTap(event: MotionEvent): Boolean {
            Log.d(TAG, "onDoubleTapEvent: tap " + event.rawX + " : " + playerView.width / 2.0)
            doubleTap = false
            doubleClickArea!!.performClick()
            if (event.rawX > playerView.width / 2.0) {
                exoFfwd!!.performClick()
            } else {
                exoRew!!.performClick()
            }
            return super.onDoubleTap(event)
        }
    }

    companion object {
        fun start(activity: Activity, serial: Serial?) {
            val intent = Intent(activity, VideoActivity::class.java)
            if (serial != null) intent.putExtra("SERIAL", serial)
            activity.startActivity(intent)
        }
    }

    class DialogFragment(private val callback: DialogCallback, var items: List<String>) : BottomSheetDialogFragment() {

        private lateinit var recycleView: RecyclerView

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            setStyle(STYLE_NORMAL, R.style.CustomBottomSheetDialogTheme)
        }

        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            val dialog = super.onCreateDialog(savedInstanceState) as BottomSheetDialog

            dialog.setOnShowListener {
                val bottomSheet: FrameLayout = dialog.findViewById(com.google.android.material.R.id.design_bottom_sheet)!!
                val behavior: BottomSheetBehavior<*> = BottomSheetBehavior.from<View>(bottomSheet)
                behavior.skipCollapsed = true
                behavior.state = BottomSheetBehavior.STATE_EXPANDED
            }
            return dialog
        }

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return inflater.inflate(R.layout.bottom_sheet_dialog, container, false)
        }

        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)
            recycleView = view.findViewById(R.id.bottomSheetRV)
            val translationAdapter = DialogRVAdapter(object : OnItemSelectListener {
                override fun onSelect(position: Int) {
                    callback.onSelect(position)
                    dismiss()
                }
            })
            translationAdapter.items = items
            recycleView.adapter = translationAdapter

        }
    }

    interface DialogCallback {
        fun onSelect(position: Int)
    }
}